lib/vrl/stdlib/src/parse_csv.rs

```diff
@@ -0,0 +1,127 @@
+use bytes::Buf;
+use csv::ReaderBuilder;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseCsv;
+
+impl Function for ParseCsv {
+    fn identifier(&self) -> &'static str {
+        "parse_csv"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse a single CSV formatted row",
+            source: r#"parse_csv!(s'foo,bar,"foo "", bar"')"#,
+            result: Ok(r#"["foo", "bar", "foo \", bar"]"#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let delimiter = arguments.optional("delimiter").unwrap_or(expr!(","));
+        Ok(Box::new(ParseCsvFn { value, delimiter }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "delimiter",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseCsvFn {
+    value: Box<dyn Expression>,
+    delimiter: Box<dyn Expression>,
+}
+
+impl Expression for ParseCsvFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let csv_string = self.value.resolve(ctx)?.try_bytes()?;
+        let delimiter = self.delimiter.resolve(ctx)?.try_bytes()?;
+        if delimiter.len() != 1 {
+            return Err("delimiter must be a single character".into());
+        }
+        let delimiter = delimiter[0];
+
+        let reader = ReaderBuilder::new()
+            .has_headers(false)
+            .delimiter(delimiter)
+            .from_reader(csv_string.bytes());
+
+        let result = reader
+            .into_byte_records()
+            .next()
+            .transpose()
+            .map_err(|err| format!("invalid csv record: {}", err))? // shouldn't really happen
+            .map(|record| record.iter().map(Into::into).collect::<Vec<Value>>())
+            .unwrap_or_default()
+            .into();
+
+        Ok(result)
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        return TypeDef::new().fallible().array::<Kind>(type_def());
+    }
+}
+
+fn type_def() -> Vec<Kind> {
+    vec![Kind::Bytes]
+}
```

Comment:
Just a note, but nothing that requires changing,

If you have a function like this for consistency or whatever reason, but it's mostly a wrapper function, you can use the `#[inline]` attribute to inform the compiler to inline this code at the call-site if it thinks that can improve performance (basically removing an extra layer of indirection through a function pointer).

There's also `#[inline(always)]`, but in general you should stick to `#[inline]` and let the compiler decide when to inline and when not to.

Reply:
Good to know, thanks! Updated.